File: cpp/searchinrotatedsortedarray.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int i = 0, j = nums.size()-1;
        int mid;
        while (i < j) {
            mid = i + (j-i)/2;
            if (nums[mid] <= nums[j]) j = mid;
            else i = mid + 1;
        }
        int k = i;
        i = 0;
        j = nums.size()-1;
        int n = nums.size();
        while (i <= j) {
            mid = i + (j-i)/2;
            if (nums[(mid+k)%n] == target) return (mid+k)%n;
            else if (nums[(mid+k)%n] < target) i = mid + 1;
            else j = mid - 1;
        }
        return -1;
    }
};

File: python/searchinrotatedsortedarray.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        # find the number of rotation
        l = 0
        h = len(nums)-1
        while l < h:
            mid = (l+h)/2
            if nums[mid] > nums[h]:
                l = mid+1
            else:
                h = mid
        
        rot = l
        l = 0
        n = len(nums)
        h = len(nums)-1
        while l <= h:
            mid = (l+h)/2
            if nums[(mid+rot)%n] == target:
                return (mid+rot)%n
            elif nums[(mid+rot)%n] > target:
                h = mid-1
            else:
                l = mid+1
        return -1
